Source/WebCore/rendering/GridMasonryLayout.cpp

```diff
@@ -0,0 +1,145 @@
+    m_masonryAxisGridGap = m_renderGrid.gridGap(m_masonryAxisDirection);
+    m_gridAxisTracksCount = m_renderGrid.numTracks(m_gridAxisDirection);
+    resizeAndResetRunningPositions(); 
+
+
+    // 2.4 Masonry Layout Algorithm
+    addItemsToFirstTrack(firstTrackItems);
+    placeItemsWithDefiniteGridAxisPosition(itemsWithDefiniteGridAxisPosition);
+    placeItemsWithIndefiniteGridAxisPosition(itemsWithIndefinitePosition);
+}
+
+void GridMasonryLayout::resizeAndResetRunningPositions()
+{
+    m_runningPositions.resize(m_gridAxisTracksCount);
+    m_runningPositions.fill(LayoutUnit());
+}
+
+void GridMasonryLayout::addItemsToFirstTrack(const HashMap<RenderBox*, GridArea>& firstTrackItems)
+{
+    for (auto& [item, gridArea] : firstTrackItems)
+        insertIntoGridAndLayoutItem(item, gridArea);
+}
+
+void GridMasonryLayout::placeItemsWithDefiniteGridAxisPosition(const Vector<RenderBox*>& itemsWithDefiniteGridAxisPosition)
+{
+    for (auto* item : itemsWithDefiniteGridAxisPosition) {
+        auto itemSpan = m_renderGrid.currentGrid().gridItemSpan(*item, m_gridAxisDirection);
+
+        ASSERT(!itemSpan.isIndefinite());
+
+        itemSpan.translate(m_renderGrid.currentGrid().explicitGridStart(m_gridAxisDirection));
+
+        GridArea area = m_masonryAxisDirection == ForRows ? GridArea(m_masonryAxisSpan, itemSpan) : GridArea(itemSpan, m_masonryAxisSpan);
+        insertIntoGridAndLayoutItem(item, area);
+    }
+}
+
+void GridMasonryLayout::placeItemsWithIndefiniteGridAxisPosition(const Vector<RenderBox*>& itemsWithIndefinitePosition)
+{
+    for (auto* item : itemsWithIndefinitePosition)
+        insertIntoGridAndLayoutItem(item, nextMasonryPositionForItem(item));
+}
+
+void GridMasonryLayout::insertIntoGridAndLayoutItem(RenderBox* child, const GridArea& area)
+{
+    m_renderGrid.currentGrid().insert(*child, area);
+
+    if (m_gridAxisDirection == ForColumns)
+        child->setOverridingContainingBlockContentLogicalWidth(m_renderGrid.m_trackSizingAlgorithm.estimatedGridAreaBreadthForChild(*child, ForColumns));
+    else
+        child->setOverridingContainingBlockContentLogicalHeight(m_renderGrid.m_trackSizingAlgorithm.estimatedGridAreaBreadthForChild(*child, ForRows));
+    child->layoutIfNeeded();
+    updateRunningPositions(child, area);
+}
+
+void GridMasonryLayout::updateRunningPositions(const RenderBox* child, const GridArea& area)
+{
+    auto gridSpan = m_masonryAxisDirection == ForRows ? area.columns : area.rows;
+
+    ASSERT(gridSpan.startLine() < m_runningPositions.size() && gridSpan.endLine() <= m_runningPositions.size());
+    gridSpan.clamp(m_runningPositions.size());
+
+    auto previousRunningPosition = m_runningPositions[gridSpan.startLine()];
+
+    auto newRunningPosition = m_masonryAxisDirection == ForRows ? previousRunningPosition + child->logicalHeight() + child->verticalMarginExtent() : previousRunningPosition + child->logicalWidth() + child->horizontalMarginExtent();
```

Comment:
I think we need to check `GridLayoutFunctions::isOrthogonalChild` to detect children that use an orthogonal coordinate system to the grid, and swap width/height for them.

Reply:
This should be fixed in https://github.com/WebKit/WebKit/pull/6940. Can you double check the logic for this though?